File: week12/23/prob5_implement_stack_using_queues/prob5.py
from collections import deque

class MyStack:
    def __init__(self):
        self.queue1 = deque()  # Main queue
        self.queue2 = deque()  # Auxiliary queue

    def push(self, x: int) -> None:
        self.queue1.append(x)

    def pop(self) -> int:
        # Move all elements except the last to queue2
        while len(self.queue1) > 1:
            self.queue2.append(self.queue1.popleft())
        # Last element of queue1 is the top of the stack
        top_element = self.queue1.popleft()
        # Swap queue1 and queue2
        self.queue1, self.queue2 = self.queue2, self.queue1
        return top_element

    def top(self) -> int:
        # Move all elements except the last to queue2
        while len(self.queue1) > 1:
            self.queue2.append(self.queue1.popleft())
        # Last element of queue1 is the top of the stack
        top_element = self.queue1.popleft()
        self.queue2.append(top_element)  # Put it back into queue2
        # Swap queue1 and queue2
        self.queue1, self.queue2 = self.queue2, self.queue1
        return top_element

    def empty(self) -> bool:
        return not self.queue1

# Example usage
stack = MyStack()
stack.push(1)
stack.push(2)
print(stack.top())   # Output: 2
print(stack.pop())   # Output: 2
print(stack.empty()) # Output: False

File: week12/23/prob5_implement_stack_using_queues/prob5.cpp
#include <iostream>
#include <queue>
using namespace std;

class MyStack {
private:
    queue<int> queue1;
    queue<int> queue2;

public:
    MyStack() {}

    void push(int x) {
        queue1.push(x);
    }

    int pop() {
        while (queue1.size() > 1) {
            queue2.push(queue1.front());
            queue1.pop();
        }
        int topElement = queue1.front();
        queue1.pop();
        swap(queue1, queue2);
        return topElement;
    }

    int top() {
        while (queue1.size() > 1) {
            queue2.push(queue1.front());
            queue1.pop();
        }
        int topElement = queue1.front();
        queue2.push(topElement);
        queue1.pop();
        swap(queue1, queue2);
        return topElement;
    }

    bool empty() {
        return queue1.empty();
    }
};

// Example usage
int main() {
    MyStack stack;
    stack.push(1);
    stack.push(2);
    cout << stack.top() << endl;   // Output: 2
    cout << stack.pop() << endl;   // Output: 2
    cout << stack.empty() << endl; // Output: 0
    return 0;
}
